### PR TITLE
src/common : proper handling of units in `strict_iec_cast`

### PR DESCRIPTION
**Pull request by neesingh-rh**
ceph.quota.max_bytes accepting invalid values like TG, GT, KT and setting the values to first letter.

Fixes: https://tracker.ceph.com/issues/67169
Signed-off-by: Neeraj Pratap Singh <neesingh@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by batrick — CHANGES_REQUESTED** [review]
Please add tests to `src/test/strtol.cc`.

**Comment by batrick** [comment]
Please also update the tracker to: https://tracker.ceph.com/issues/67169

**Comment by neesingh-rh** [comment]
jeninks test make check

File: src/common/strtol.cc [hunk 1]
```diff
@@ -148,41 +148,47 @@ T strict_iec_cast(std::string_view str, std::string *err)
...
-      default:
-        *err = "strict_iecstrtoll: unit prefix not recognized";
-        return 0;
+    if ((unit.back() == 'i') || (unit.length() == 1)) {
+      if (unit.back() == 'i') {
+        if (unit.front() == 'B') {
+          *err = "strict_iecstrtoll: illegal prefix \"Bi\"";
+          return 0;
+        }
+      }
```
Comment (dparmar18) [inline src/common/strtol.cc:161]: IMO we need to also catch cases like `BB` or `GG`. In total three cases to handle: 
1) `i` 
2) `Bi`
3) cases like `GG` or `TT`

I have some suggestions, feel free to amend/reject/modify:

```
// catch case for prefix "i"
if (unit.length() == 1) {
	if ((unit.front() == 'i') && (unit.back() == 'i')) {
		*err = "strict_iecstrtoll: illegal prefix \"i\"";
	}
	return 0;
}

// catch case "Bi" or repetitiveness like "GG" or "TT"
if (unit.length() == 2) {
	if (unit.front() == unit.back()) {
		*err = "strict_iecstrtoll: illegal prefix "  << "\"" << unit << "\"";
	} else if ((unit.front() == 'B') && (unit.back() == 'i')) { 
		*err = "strict_iecstrtoll: illegal prefix \"Bi\"";
	}
	if (!err.empty()) {
		return 0;
	}
}
```

File: src/test/strtol.cc [hunk 2]
```diff
@@ -257,6 +257,21 @@ TEST(StrictIECCast, Error) {
     (void)strict_iec_cast<int>("1T", &err);
```
Comment (dparmar18) [inline src/test/strtol.cc:257]: Huh? `1T` is allowed right? 

```
build$ setfattr -n ceph.quota.max_bytes -v 1T ./mnt/foo/
build$ getfattr -n ceph.quota.max_bytes ./mnt/foo/
# file: mnt/foo/
ceph.quota.max_bytes="1099511627776"
```

I know this is out of scope of this PR, no worries but this something to be noted @batrick @vshankar

**Comment by dparmar18** [comment]
@neesingh-rh apart from adding test cases to `src/test/strtol.cc` (which are generic cases for func `strict_iec_cast`) you'd need to add CephFS quota specific cases too i.e. `qa/tasks/cephfs/test_quota.py` (check `test_human_readable_quota_values()`, `test_human_readable_quota_invalid_values()` and `test_disable_enable_human_readable_quota_values()`)

File: src/common/strtol.cc [hunk 3]
```diff
@@ -148,41 +148,47 @@ T strict_iec_cast(std::string_view str, std::string *err)
...
+          break;
+        case 'B':
+          break;
+        default:
+          *err = "strict_iecstrtoll: unit prefix not recognized";
+          return 0;
+      }
     }
+    else {
+      *err = "strict_iecstrtoll: illegal unit prefix";
```
Comment (dparmar18) [inline src/common/strtol.cc:189]:
```suggestion
      *err = "strict_iecstrtoll: illegal prefix "  << "\"" << unit << "\"";
```

File: src/common/strtol.cc (hunk 1 above)
Comment (dparmar18) [reply src/common/strtol.cc:161]: NVM, the switch is wrapped inside `if ((unit.back() == 'i') || (unit.length() == 1)) {`

File: src/common/strtol.cc [hunk 4]
```diff
@@ -148,41 +148,47 @@ T strict_iec_cast(std::string_view str, std::string *err)
...
-      case 'B':
-        break;
-      default:
-        *err = "strict_iecstrtoll: unit prefix not recognized";
-        return 0;
+    if ((unit.back() == 'i') || (unit.length() == 1)) {
+      if (unit.back() == 'i') {
+        if (unit.front() == 'B') {
+          *err = "strict_iecstrtoll: illegal prefix \"Bi\"";
+          return 0;
```
Comment (dparmar18) [inline src/common/strtol.cc:159]: We can handle the case of `i` (single char) here itself

```suggestion
	if (unit.front() == 'i') {
	  *err = "strict_iecstrtoll: illegal prefix \"i\"";
	} else if (unit.front() == 'B') {
	  *err = "strict_iecstrtoll: illegal prefix \"Bi\"";
	}
	if (!err.empty()) {
	  return 0;
	}
```

File: src/test/strtol.cc (hunk 2 above)
Comment (neesingh-rh) [reply src/test/strtol.cc:257]: `1T` is beyond the range of `int` type, will throw error.

**Comment by neesingh-rh** [comment]
> @neesingh-rh apart from adding test cases to `src/test/strtol.cc` (which are generic cases for func `strict_iec_cast`) you'd need to add CephFS quota specific cases too i.e. `qa/tasks/cephfs/test_quota.py` (check `test_human_readable_quota_values()`, `test_human_readable_quota_invalid_values()` and `test_disable_enable_human_readable_quota_values()`)

As this PR is related to non-acceptance of invalid values, I have added some more invalid values in `test_human_readable_quota_invalid_values()`

File: src/common/strtol.cc (hunk 4 above)
Comment (neesingh-rh) [reply src/common/strtol.cc:159]: I don't think we do need this as `Bi` is a exception from rest of the prefixes( As `B` is valid but `Bi` is not), it needs to be handled separately. And rest of the cases are anyways handled by the switch statement.

File: src/test/strtol.cc (hunk 2 above)
Comment (dparmar18) [reply src/test/strtol.cc:257]: Ah gotcha, the used cast type is `int` there.

File: src/common/strtol.cc (hunk 4 above)
Comment (dparmar18) [reply src/common/strtol.cc:159]: with the current code `i` being an invalid type would be evaluated in switch and would say `strict_iecstrtoll: unit prefix not recognized`. It should speak of `i` for clarity. Please append `unit` to the statement. (at line 184)

File: qa/tasks/cephfs/test_quota.py [hunk 5]
```diff
@@ -135,7 +135,8 @@ def test_human_readable_quota_invalid_values(self):
 
         self.mount_a.run_shell(["mkdir", "subdir"])
 
-        invalid_values = ["10A", "1y00Ki", "af00", "G", "", " ", "-1t", "-1"]
+        invalid_values = ["10A", "1y00Ki", "af00", "G", "", " ", "-1t", "-1",
+                          "1GT", "2MM", "5Di", "8Bi"]
```
Comment (dparmar18) [inline qa/tasks/cephfs/test_quota.py:139]: Please add cases like `i`, `3iB`

File: src/common/strtol.cc (hunk 4 above)
Comment (neesingh-rh) [reply src/common/strtol.cc:159]: Appending unit to the `err` string makes sense.

File: qa/tasks/cephfs/test_quota.py (hunk 5 above)
Comment (neesingh-rh) [reply qa/tasks/cephfs/test_quota.py]: Added.

File: src/common/strtol.cc [hunk 6]
```diff
@@ -148,41 +148,51 @@ T strict_iec_cast(std::string_view str, std::string *err)
...
+          break;
+        case 'E':
+          m = 60;
+          break;
+        case 'B':
+          break;
+        default:
+          ostringstream errStr;    
+          errStr << "strict_iecstrtoll: unit prefix not recognized '"  << unit << "' ";
+          *err = errStr.str();
```
Comment (dparmar18) [inline src/common/strtol.cc:186]: An additional storage? Is this really needed?

Comment (neesingh-rh) [reply src/common/strtol.cc]: Addressed

File: src/common/strtol.cc [hunk 7]
```diff
@@ -148,41 +148,47 @@ T strict_iec_cast(std::string_view str, std::string *err)
     unit = str.substr(u, str.length() - u);
     // we accept both old si prefixes as well as the proper iec prefixes
     // i.e. K, M, ... and Ki, Mi, ...
-    if (unit.back() == 'i') {
-      if (unit.front() == 'B') {
-        *err = "strict_iecstrtoll: illegal prefix \"Bi\"";
-        return 0;
-      }
-    }
     if (unit.length() > 2) {
```
Comment (batrick) [inline src/common/strtol.cc:151]: What about `1KiB`? (where `unit.length() == 3`)?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/common/strtol.cc (hunk 7 above)
Comment (neesingh-rh) [reply src/common/strtol.cc:156]: handled cases like KB, MB and KiB, MiB etc.

File: src/test/strtol.cc [hunk 8]
```diff
@@ -257,6 +257,21 @@ TEST(StrictIECCast, Error) {
...
+  {
+    std::string err;
+    (void)strict_iec_cast<int64_t>("1TG", &err);
+    ASSERT_NE(err, "");
+  }
+  {
+    std::string err;
+    (void)strict_iec_cast<int64_t>("1KD", &err);
+    ASSERT_NE(err, "");
+  }
```
Comment (batrick) [inline src/test/strtol.cc:274]: Please add tests for `1KiB`/etc.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/test/strtol.cc (hunk 8 above)
Comment (neesingh-rh) [reply src/test/strtol.cc:274]: Updated.

**Comment by dparmar18** [comment]
please align the PR/commit title to the issue being fixed

File: src/test/strtol.cc (hunk 8 above)
Comment (batrick) [reply src/test/strtol.cc:274]: It occurred to me now that this test case is for checking erroneous units. Please move these tests to

`TEST(IECStrToLL, WithUnits)`

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/test/strtol.cc (hunk 8 above)
Comment (neesingh-rh) [reply src/test/strtol.cc:286]: Yup, you are right!

**Review by batrick — APPROVED** [review]

**Review by dparmar18 — APPROVED** [review]

**Comment by neesingh-rh** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68395](https://tracker.ceph.com/issues/68395).

**Comment by rishabh-d-dave** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68419](https://tracker.ceph.com/issues/68419).

**Review by rishabh-d-dave — APPROVED** [review]
QA run was successful - https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-rishabh-testing-20241007102024

**Pull request merged**